internal/wasm/memory.go

```diff
@@ -267,6 +290,22 @@ func PagesToUnitOfBytes(pages uint32) string {
 
 // Below are raw functions used to implement the api.Memory API:
 
+// Uses atomic to set the length of the buf slice to the maximum
+// of the previous and new lengths.
+func atomicGrowLength(buf *[]byte, length uintptr) {
+	slicePtr := (*reflect.SliceHeader)(unsafe.Pointer(buf))
+	lenPtr := (*uintptr)(unsafe.Pointer(&slicePtr.Len))
+	for {
+		if old := atomic.LoadUintptr(lenPtr); old >= length {
+			// Was already greater or equal.
+			break
+		} else if atomic.CompareAndSwapUintptr(lenPtr, old, length) {
+			// Successfully updated.
+			break
+		}
+	}
+}
```

Comment:
definitely need test

Reply:
@anuraaga convinced me to revert this, but will add a test for the new version anyway.